Add support for external IAM Instance Profiles

rifelpet opened the pull request:
This extends #3597 and pulls in some work from #2440

I confirmed that the LC contains the desired instance profile using direct, terraform, and cloudformation targets.

Thank you @chrislovecnm for doing most of the work 👍

chrislovecnm commented on pkg/model/iam.go:
```diff
@@ -54,6 +54,13 @@ func (b *IAMModelBuilder) Build(c *fi.ModelBuilderContext) error {
 	// Collect the roles in use
 	var roles []kops.InstanceGroupRole
 	for _, ig := range b.InstanceGroups {
+		if ig.Spec.IAM != nil && ig.Spec.IAM.Profile != nil {
```
So using this pattern where we do not create the tasks bypasses our validation for shared objects.  We still need to create the tasks but mark the tasks as shared. Take a look at how I am doing that in https://github.com/kubernetes/kops/pull/3625.  Please ping me if you have questions.  We will need to create shared roles, profiles, etc.  Also, delete kind bugs me since we will not have shared tags, please let me know how delete is working.

Looks great, we are really close.

rifelpet replied:
The `buildShareIAMInstanceProfile` function is already creating an `IAMInstanceProfile` task marked as shared. You're saying I need to also create a shared `IAMInstanceProfileRole` task? I can get that added, as well as test deletion behavior.

Also, do you think its worth it to add a validation warning/error when both `additionalPolicies` and `.iam.profile` are set?

chrislovecnm replied:
Yes we need the role and policy tasks. @justinsb is pretty passionate about that. 

In regards to the delete. Kops will try to delete the profile of the name is the same as what kops typically creates. I think that is a limitation we will need to live with.

rifelpet replied:
One more scenario I’m unsure about: if an existing cluster is updated to use this feature, will the original roles ever get deleted? The first ‘update cluster’ can’t delete them because they’ll be in use until a ‘rolling update’ has been ran at which point Kops won’t even know about the original roles. 

Shouldn’t deletion reference the cluster and IG specs, determine the roles are shared, and not to look for roles based off the original naming scheme?

chrislovecnm commented:
/ok-to-test

rifelpet commented on pkg/model/iam.go:
```diff
@@ -182,6 +205,64 @@ func (b *IAMModelBuilder) Build(c *fi.ModelBuilderContext) error {
 }
 
+// buildSharedIAMTasks adds a shared IAM tasks to the ModelBuilderContext
+func (b *IAMModelBuilder) buildSharedIAMTasks(id *string, c *fi.ModelBuilderContext) error {
```
@chrislovecnm I've updated this to create the additional IAM tasks as well.

Since theres a lot of duplicate logic between this and the regular IAM task creation, I considered making this a function that accepts a `shared bool` and a role name (either `b.IAMName(role)` if managed or `findCustomAuthNameFromArn(roleArn)` if shared). This would create these shared tasks with the same policy documents and such that the managed tasks would have. I'm not sure if thats desirable or if we'd just be ignoring those values in the `Find` function of each shared task.

Thoughts?

chrislovecnm commented:
/assign @rdrgmnzs

chrislovecnm commented:
/assign

chrislovecnm commented:
BTW - you are probably going to need to use phases for testing and probably need a PR that I have.

```
kops update cluster $CLUSTER --phase network
kops update cluster $CLUSTER --phase security
kops update cluster $CLUSTER --phase cluster --phase-overrides "security=WarnIfInsufficientAccess" --yes
```

That will allow the roles to be ignored :)

The PR is #4153

chrislovecnm commented:
@rifelpet this closes a couple of open issues.  Adding a note for you or me to find them.

rifelpet commented:
This closes #1166 and partially addresses #1644 and #4124

rdrgmnzs commented on pkg/model/iam.go:
```diff
@@ -51,134 +51,161 @@ const RolePolicyTemplate = `{
-	// Generate IAM objects etc for each role
-	for _, role := range roles {
-		name := b.IAMName(role)
+	// Generate IAM tasks for each managed role
```
Running a test on this it seems there's a logic bug somewhere. If you have more than 1 Instance Group using the same IAM role it errors out. This is important since in a HA setup each master will have it's own "Instance Group" and will share an IAM role. People may also choose to have multiple node Instance Groups sharing the same IAM role.

`task.go:57] found duplicate tasks with name "IAMRole/KubernetesNodeRole": *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"KubernetesNodeRole","RolePolicyDocument":{"Name":"","Resource":{}},"Shared":true,"ExportWithID":"nodes"} and *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"KubernetesNodeRole","RolePolicyDocument":{"Name":"","Resource":{}},"Shared":true,"ExportWithID":"nodes"}`

rifelpet replied:
Strange, it should only error if the same instance profile is used for multiple Instance Group Roles (master and node, node and bastion, etc.). Multiple instance groups using the same instance profile shouldn't fail if they are the same IGR. I'll take a closer look and let you know when I have a fix.

rdrgmnzs replied:
Tested this and looks like it's fixed.

kesor replied:
> Tested this and looks like it's fixed.

This is still an issue in v1.17 and 1.18 and has not been fixed :(
Opened an issue no this #9852

rdrgmnzs commented on tests/integration/update_cluster/existing_iam/in-v1alpha2.yaml:
```diff
@@ -0,0 +1,84 @@
+apiVersion: kops/v1alpha2
```
Could we make this test a master HA test, similar to what we find in https://github.com/kubernetes/kops/blob/master/tests/integration/update_cluster/ha/in-v1alpha1.yaml. This would help catch any future issues changes to this code may cause to HA setups.

rifelpet replied:
Good idea, will do.

rifelpet replied:
Test updated to use HA masters

rdrgmnzs replied:
Looks like you only have 2 masters in the test, that would actually fail as Kops requires you to have a odd number of masters for an HA setup.

rifelpet commented on pkg/apis/kops/instancegroup.go:
```diff
@@ -114,6 +114,8 @@ type InstanceGroupSpec struct {
 	// AdditionalUserData is any aditional user-data to be passed to the host
 	AdditionalUserData []UserData `json:"additionalUserData,omitempty"`
+	// IAMProfileSpec defines the identity of the cloud group iam profile.
+	IAM *IAMProfileSpec `json:"iam,omitempty"`
```
This PR includes the commit from #3597. @chrislovecnm will need to confirm but I believe the plan is that once this is approved we'll merge #3597, then rebase and merge this.

rifelpet commented:
/retest

rdrgmnzs commented on tests/integration/update_cluster/existing_iam/in-v1alpha2.yaml:
```diff
@@ -64,6 +68,26 @@ spec:
 
 ---
 
+apiVersion: kops/v1alpha2
```
This would actually cause the test to fail, Kops requires an odd number of masters. So when testing an HA setup we;d need at least 3 masters defined.

rdrgmnzs commented:
Things are looking good, we should be good to land once the tests have been fixed.

rifelpet commented:
Added a 3rd master to the HA test. looks like e2e is still failing, I'm assuming this is the unrelated e2e issue we've been experiencing recently.

rdrgmnzs commented:
/retest

rdrgmnzs commented:
/lgtm

rifelpet commented:
I had to update some tests to match master but it should be ready to go now 👍

rifelpet commented:
Ping @chrislovecnm @rdrgmnzs

rdrgmnzs commented:
@rifelpet we'd like @justinsb to take a look as well before merging.

chrislovecnm commented:
@rifelpet when you rebase can you squash?

chrislovecnm commented:
@rdrgmnzs you still have this PR marked as requesting changes, but you did a lgtm, so I assume that we are good to go.

rifelpet commented:
@chrislovecnm I've squashed the commits

justinsb commented on upup/pkg/fi/cloudup/awstasks/iaminstanceprofile.go:
```diff
@@ -93,17 +95,28 @@ func (e *IAMInstanceProfile) Run(c *fi.Context) error {
 
-func (s *IAMInstanceProfile) CheckChanges(a, e, changes *IAMInstanceProfile) error {
+func (_ *IAMInstanceProfile) ShouldCreate(a, e, changes *IAMInstanceProfile) (bool, error) {
+	if fi.BoolValue(e.Shared) {
```
Instead of doing this, I think we should use the lifecycle options.

We still want to validate that the IAM role matches expectations, or else we have no way of verifying that it is correct.  So where we would set shared == true, we probably want Lifecycle to be LifecycleExistsAndValidates.

The tricky thing is it's probably going to make overriding it harder - I'm guessing we want some sort of oracle function for lifecycle cc @chrislovecnm

chrislovecnm replied:
@rifelpet take a look at https://github.com/kubernetes/kops/pull/4445
This will allow a user to set the lifecycle on whichever tasks they want to.  

@justinsb just thought of an edge case for this.  With this feature, you probably want to set the task`IAMRolePolicy` with an overridden lifecycle. I am thinking this may stop a user from having additional policies attached.  

I am thinking that https://github.com/rifelpet/kops/blob/12f72d405e06e47a74c82e60dd6b761253d60722/pkg/apis/kops/cluster.go#L160 will no longer work, which you override the lifecycle.

@rifelpet is what I am thinking, make any sense?

This edge case is fine since you can just define what you want for the IAMProfile, but we need to document this.  Not certain how my security group PR will act.

rifelpet replied:
@justinsb with IAM-related resources there are many ways they could be setup but still function as kops would expect. inline vs managed policies, a superset of expected policies, even inline policies named differently. How would the validation handle these scenarios? Are we sure its feasible to perform this validation effectively?

In my specific case, I will be using managed policies and attaching them to the roles. Would the validation step play nicely with that?

@chrislovecnm does #3453 implement the lifecycle changes based on #4445 ? If so I can mimic the similar changes here, otherwise I'm not sure exactly what needs to happen in the code to accomplish what you're describing. (Not too familiar with the relationship between lifecycles and tasks, but ill take a closer look hopefully later today)

rifelpet replied:
Oh, unless you're implying that validation simply checks for the existence of the instance profile. That makes more sense. Do we have other tasks that perform validation on LifecycleExistsAndValidates that I could use as an example? I'm not seeing any, but basically we'll just want to make sure `findIAMInstanceProfile` doesnt return nil. Is that correct?

justinsb commented on upup/pkg/fi/cloudup/awstasks/iaminstanceprofilerole.go:
```diff
@@ -117,6 +119,13 @@ type terraformIAMInstanceProfile struct {
+	shared := fi.BoolValue(e.Shared)
+	if shared {
+		// Not terraform owned / managed
+		// We won't apply changes, but our validation (kops update) will still warn
```
Ah - this type is good.  We do need Shared for terraform / cloudformation, but we do want validation.

justinsb commented on upup/pkg/fi/cloudup/awstasks/iamrole.go:
```diff
@@ -121,6 +123,10 @@ func (e *IAMRole) Run(c *fi.Context) error {
 }
 
+func (_ *IAMRole) ShouldCreate(a, e, changes *IAMRole) (bool, error) {
+	return !fi.BoolValue(e.Shared), nil
```
Would be happier if this followed the same pattern as IAMInstanceProfileRole

chrislovecnm commented:
@rifelpet I rebased https://github.com/kubernetes/kops/pull/3597 - if you have questions about using the phase overrides please reach out

chrislovecnm commented on pkg/model/iam.go:
```diff
@@ -103,7 +108,7 @@ func (b *IAMModelBuilder) buildIAMTasks(igRole kops.InstanceGroupRole, iamName s
 		iamRole = &awstasks.IAMRole{
 			Name:      s(iamName),
-			Lifecycle: b.Lifecycle,
+			Lifecycle: &subtaskLifecycle,
```
This should be the regular lifecycle for the security phase. Not certain why you are making a new one.

chrislovecnm commented on cmd/kops/integration_test.go:
```diff
@@ -338,20 +344,35 @@ func runTestAWS(t *testing.T, clusterName string, srcDir string, version string,
-	if private {
+	manageIAM := true
+	for _, override := range lifecycleOverrides {
+		if override == "IAMRole=Ignore" {
```
These should be a list of

Task1=Ignore,Task2=Ignore.

You will need to split the sting on commas.

chrislovecnm commented on docs/iam_roles.md:
````diff
@@ -139,3 +141,40 @@ You can have an additional policy for each kops role (node, master, bastion). Fo
+Now run a cluster update to create the new launch configuration:
+
+```
+kops update cluster ${CLUSTER_NAME} --yes
````
Need note about overriding the lifecycle.

chrislovecnm commented on pkg/model/iam.go:
```diff
@@ -51,133 +51,167 @@ const RolePolicyTemplate = `{
+			Name:      s(iamName),
+			Lifecycle: &subtaskLifecycle,
+			Shared:    fi.Bool(shared),
+
```
Do we need shared here? Not certain which task is the instance profile.

chrislovecnm commented on pkg/model/iam.go:
```diff
@@ -51,133 +51,167 @@ const RolePolicyTemplate = `{
+		iamInstanceProfile = &awstasks.IAMInstanceProfile{
+			Name:      s(iamName),
+			Lifecycle: b.Lifecycle,
+			Shared:    fi.Bool(shared),
```
I think this is the only shared task we need.

chrislovecnm replied:
Let me be more clear. We should not need shared on the other tasks.

chrislovecnm commented on upup/pkg/fi/cloudup/awstasks/iaminstanceprofile.go:
```diff
@@ -93,17 +95,28 @@ func (e *IAMInstanceProfile) Run(c *fi.Context) error {
 }
 
-func (s *IAMInstanceProfile) CheckChanges(a, e, changes *IAMInstanceProfile) error {
+func (_ *IAMInstanceProfile) ShouldCreate(a, e, changes *IAMInstanceProfile) (bool, error) {
```
Can we remove?

rifelpet commented:
Updated to use lifecycle overrides rather than `Shared`, and updated docs to match. Ready for review @justinsb @chrislovecnm @rdrgmnzs

I didn't squash so you can see just the latest changes, but I can squash whenever we're ready to merge.

chrislovecnm commented on pkg/model/iam.go:
```diff
@@ -51,133 +51,158 @@ const RolePolicyTemplate = `{
+	{
+		additionalPolicy := ""
+		if b.Cluster.Spec.AdditionalPolicies != nil {
+			roleAsString := reflect.ValueOf(igRole).String()
```
I know this is not your code, so let's not fix it in this PR.  But using reflection for this looks ugly.  We can refactor later but I think we can just do `string(igRole)`.  Am I correct?  If so I will file an issue.

chrislovecnm commented on upup/pkg/fi/cloudup/awstasks/iaminstanceprofile.go:
```diff
@@ -93,17 +95,21 @@ func (e *IAMInstanceProfile) Run(c *fi.Context) error {
 }
 
-func (s *IAMInstanceProfile) CheckChanges(a, e, changes *IAMInstanceProfile) error {
+func (_ *IAMInstanceProfile) CheckChanges(a, e, changes *IAMInstanceProfile) error {
```
Why did we change the receiver?

chrislovecnm replied:
It may be incorrect, but all of the other IAM tasks have `func (s *IAMInstanceProfile)`

rifelpet replied:
It might've been my IDE complaining. I just reverted this change

rdrgmnzs commented on docs/iam_roles.md:
````diff
@@ -139,3 +141,45 @@ You can have an additional policy for each kops role (node, master, bastion). Fo
+Now run a cluster update to create the new launch configuration, using [lifecycle overrides](./cli/kops_update_cluster.md#options) to prevent IAM-related resources from being created:
+
+```
+kops update cluster ${CLUSTER_NAME} --yes --lifecycle-overrides "IAMRole=Ignore,IAMRolePolicy=Ignore,IAMInstanceProfileRole=Ignore"
````
When testing having quotes around --lifecycle-overrides "IAMRole=Ignore,IAMRolePolicy=Ignore,IAMInstanceProfileRole=Ignore" makes it fail. Removing the quotes fixed this.

rdrgmnzs commented:
@chrislovecnm @rifelpet  any reasons we are using the full ARN in the Kops spec? Looks like we're extracting just the IAM name from the ARN to use anyway no? Wouldn't just using the IAM name simplify things?

Otherwise things look good.

rifelpet commented:
@rdrgmnzs I agree using just the name would simply things. It might help this field become cloud-agnostic down the road as well. With @chrislovecnm 's go-ahead I can strip the arn logic out of this and use just the instance profile name

chrislovecnm commented:
Did I not respond to this ... opps.

1. Do not care about the name much as long as it works
2. I need to be able to use names or arns with the paths in them

rarescosma commented:
Bump

mad01 commented:
what is the state of this?

rifelpet commented:
@chrislovecnm  @rdrgmnzs It looks like the [AWS API](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_CreateLaunchConfiguration.html) accepts either ARN or name. We could probably drop the ARN regex validation to allow either, then either use the name or strip the name out of the ARN to use for the awstask name.

This could also happen in a later PR as it would only be adding additional functionality. Otherwise this should be good to merge. Thoughts?

justinsb commented on pkg/model/iam.go:
```diff
@@ -51,133 +51,158 @@ const RolePolicyTemplate = `{
-				Lifecycle: b.Lifecycle,
+func (b *IAMModelBuilder) buildIAMTasks(igRole kops.InstanceGroupRole, iamName string, c *fi.ModelBuilderContext, shared bool) error {
+	var iamRole *awstasks.IAMRole
+	{
```
Can we insert another level of indentation here - the indentation change makes the diff hard to review.

e.g.

```
{ // To minimize diff for easier code review
  { // To minimize diff for easier code review
...
```

justinsb commented on tests/integration/update_cluster/existing_iam/kubernetes.tf:
```diff
@@ -0,0 +1,533 @@
+  image_id                    = "ami-15000000"
+  instance_type               = "m3.medium"
+  key_name                    = "${aws_key_pair.kubernetes-existing-iam-example-com-c4a6ed9aa889b9e2c39cd663eb9c7157.id}"
+  iam_instance_profile        = "kops-custom-master-role"
```
I don't see where this is defined.  It's not documented in the TF provider, but I'm guessing this should be the ARN https://www.terraform.io/docs/providers/aws/r/launch_configuration.html#iam_instance_profile

I did notice that the shared tasks didn't appear to use the ARN, so I think this is a genuine issue - I don't think it just applies to terraform.

justinsb commented on pkg/model/names.go:
```diff
@@ -114,9 +117,28 @@ func (b *KopsModelContext) IAMName(role kops.InstanceGroupRole) string {
+func (b *KopsModelContext) LinkToIAMInstanceProfile(ig *kops.InstanceGroup) (*awstasks.IAMInstanceProfile, error) {
+	if ig.Spec.IAM != nil && ig.Spec.IAM.Profile != nil {
+		name, err := findCustomAuthNameFromArn(fi.StringValue(ig.Spec.IAM.Profile))
+		return &awstasks.IAMInstanceProfile{Name: &name}, err
```
I think we're going to need to pass the ARN in here

rifelpet commented on cmd/kops/integration_test.go:
```diff
@@ -71,81 +71,92 @@ func TestHighAvailabilityGCE(t *testing.T) {
+
+// TestExistingIAMCloudformation runs the test with existing IAM instance profiles, similar to kops create cluster minimal.example.com --zones us-west-1a
+func TestExistingIAMCloudformation(t *testing.T) {
+	lifecycleOverrides := []string{"IAMRole=Ignore", "IAMRolePolicy=Ignore", "IAMInstanceProfileRole=Ignore"}
```
Would `LifecycleExistsAndWarnIfChanges` allow us to use managed policies as opposed to inline policies? The role has identical permissions but the IAM resources will not be in the state that the awstasks are expecting. In our case we use `iam:AttachRolePolicy` rather than `iam:PutRolePolicy`, so the `IAMRolePolicy` awstask would see that the inline policy is missing.

rifelpet commented:
@justinsb I've added indentation so the diff is more readable. Regarding the docs update, can you explain how exactly kops would check for differences in policy permissions if we're allowing users to define permissions in a variety of ways? inline vs managed policies, etc.

I'm not sure how kops could diff this. It might be better to just rely on explicitly mentioning policy changes in release notes, in which case I can add a note in the docs to check them.

justinsb commented:
So on `Ignore` vs `LifecycleExistsAndWarnIfChanges` - we could use `LifecycleExistsAndValidates` which will fail if the IAM tasks are different, but because we do expect that IAM policies won't be exactly identical I was suggesting `LifecycleExistsAndWarnIfChanges`.  The idea being that we then check that we accept that you don't have the identical policy, but we highlight it so that when someone says their cluster isn't working it should be obvious why that is.  If we do `Ignore` it's much harder to see.

It might well be that LifecycleExistsAndWarnIfChanges isn't working as I expect / for your configuration though!  Is that the case?

rifelpet commented:
In my case `ExistsAndWarnIfChanges` does not break but I have a feeling its not behaving as one might expect.

I have managed policies attached to the roles, whereas kops creates inline policies. I would expect to run into issues with any `Exists*` Lifecycle override because the IAMRolePolicy awstask does not (and should not) exist since it refers to an inline policy.

However, using `ExistsAndWarnIfChanges` results in the following output:
```
I0510 15:29:49.072687   20915 executor.go:91] Tasks: 0 done / 91 total; 38 can run
Object from different phase did not match, problems possible:
  IAMRole/?
  	RolePolicyDocument
  	                    	+ {
  	                    	+   "Version": "2012-10-17",
  	                    	+   "Statement": [
  	                    	+     {
  	                    	+       "Effect": "Allow",
  	                    	+       "Principal": { "Service": "ec2.amazonaws.com"},
  	                    	+       "Action": "sts:AssumeRole"
  	                    	+     }
  	                    	+   ]
  	                    	- {"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"ec2.amazonaws.com","AWS":"arn:aws:iam::123456789012:root"},"Action":"sts:AssumeRole"}]}
  	                    	+ }


W0510 15:29:49.612410   20915 executor.go:110] Lifecycle set to ExistsAndWarnIfChanges and object was not found.
I0510 15:29:49.612474   20915 executor.go:91] Tasks: 38 done / 91 total; 25 can run
W0510 15:29:50.072392   20915 executor.go:110] Lifecycle set to ExistsAndWarnIfChanges and object was not found.
W0510 15:29:50.072422   20915 executor.go:110] Lifecycle set to ExistsAndWarnIfChanges and object was not found.
W0510 15:29:50.072433   20915 executor.go:110] Lifecycle set to ExistsAndWarnIfChanges and object was not found.
W0510 15:29:50.072442   20915 executor.go:110] Lifecycle set to ExistsAndWarnIfChanges and object was not found.
W0510 15:29:50.072452   20915 executor.go:110] Lifecycle set to ExistsAndWarnIfChanges and object was not found.
...
```

The role policy document update that is being reported is fine for me to ignore, however the warnings that follow I would expect to be errors that should fail validation. It'd be nice if they reported the task type and name to provide additional clarity.

It seems like something in the stacktrace to [this](https://github.com/kubernetes/kops/blob/master/upup/pkg/fi/context.go#L117) function should be treating its error handling differently.

I'll update the docs and tests to use `ExistsAndWarnIfChanges` though we should probably fix this behavior before the next release. WDYT?
@justinsb @chrislovecnm

justinsb commented:
Thanks everyone for bearing with the process here!

/approve
/lgtm

gerr1t commented:
@rifelpet , thanks a lot, have been waiting for this one.

kplimack commented:
I'm failing on this on `1.11.0`
```
F0311 11:18:54.393929   65298 task.go:59] found duplicate tasks with name "IAMRole/nodes.k8s-development.myorg.com": *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"nodes.k8s-development.myorg.com","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"} and *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"nodes.k8s-development.sendgrid.net","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"}
```

kesor commented:
Failing on this as well in `1.18`

```
$ kops update cluster
I0830 22:25:43.874299   86529 dns.go:95] Private DNS: skipping DNS validation
F0830 22:25:43.921974   86529 task.go:59] found duplicate tasks with name "IAMRole/nodes.myk8sname": *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"nodes.myk8sname","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"} and *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"nodes.myk8sname","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"}
```

rifelpet commented:
Please open a bug report issue that includes your full cluster and instance groups specs.
